src/main/java/com/github/manolo8/darkbot/extensions/plugins/PluginUpdater.java

```diff
@@ -0,0 +1,288 @@
+
+        private final StringBuilder failedUpdates = new StringBuilder();
+
+        private int progress = 0;
+        private final List<UpdateTask> updateTasks = new ArrayList<>();
+
+        UpdateAllTask() {
+            List<Plugin> availableUpdates = pluginHandler.getAvailableUpdates().collect(Collectors.toList());
+            availableUpdates.forEach(pl -> updateTasks.add(new UpdateTask(pl, this)));
+            mainProgressBar.setMaximum(availableUpdates.size() * 5 + 1);
+        }
+
+        public synchronized void tickUpdate() {
+            super.publish(progress += 1);
+        }
+
+        @Override
+        protected void process(List<Integer> chunks) {
+            mainProgressBar.setValue(chunks.get(chunks.size() - 1));
+        }
+
+        @Override
+        protected void done() {
+            mainProgressBar.setValue(mainProgressBar.getMaximum());
+            if (failedUpdates.length() > 0)
+                Popups.showMessageAsync("Update failed", failedUpdates, JOptionPane.ERROR_MESSAGE);
+        }
+
+        @Override
+        protected Void doInBackground() {
+            updateTasks.forEach(SwingWorker::execute);
+            updateTasks.forEach(UpdateTask::waitUntilDone);
+
+            pluginHandler.updatePlugins();
+
+            updateTasks.forEach(task -> {
+                if (task.failed)
+                   failedUpdates.append(task.plugin.getName()).append(" failed to update\n");
+            });
+            return null;
+        }
+    }
+
+    private class UpdateTask extends SwingWorker<Void, Void> {
+
+        private final StringBuilder failedUpdate = new StringBuilder();
+
+        private final Plugin plugin;
+        private final UpdateAllTask updateAllTask;
+        private final ProgressManager progressManager;
+
+        private boolean failed;
+
+        UpdateTask(Plugin plugin) {
+            this(plugin, null);
+        }
+
+        UpdateTask(Plugin plugin, UpdateAllTask updateAllTask) {
+            this.plugin = plugin;
+            this.updateAllTask = updateAllTask;
+            PluginCard card = pluginDisplay.getPluginCard(plugin);
+            this.progressManager = new ProgressManager(card, this);
+        }
+
+        private void waitUntilDone() {
+            try {
+                get();
+            } catch (InterruptedException | ExecutionException e) {
+                e.printStackTrace();
+            }
+        }
+
+        @Override
+        protected void done() {
+            plugin.setUpdateStatus(failed ? Plugin.UpdateStatus.FAILED : Plugin.UpdateStatus.UP_TO_DATE);
+            if (updateAllTask != null || failedUpdate.length() <= 0) return;
+
+            Popups.showMessageAsync("Update failed", failedUpdate, JOptionPane.ERROR_MESSAGE);
+        }
+
+        @Override
+        protected Void doInBackground() throws Exception {
+            progressManager.setUpdateProgress(ProgressManager.State.STARTING);
```

Comment:
You can't modify UI code from background thread. You MUST do this from the `process` method of the worker.
Call `publish(State.STARTING)`, and on process grab the latest state and push it via `card.setUpdateProgress(state)` (see the comment in the ProgressManager class)